server/commands/teamCreator.ts

```diff
@@ -0,0 +1,132 @@
+import { Transaction } from "sequelize";
+import slugify from "slugify";
+import { RESERVED_SUBDOMAINS } from "@shared/utils/domains";
+import { sequelize } from "@server/database/sequelize";
+import Logger from "@server/logging/Logger";
+import { APM } from "@server/logging/tracing";
+import { Team, Event } from "@server/models";
+import { generateAvatarUrl } from "@server/utils/avatars";
+
+type Props = {
+  /** The displayed name of the team */
+  name: string;
+  /** The domain name from the email of the user logging in */
+  domain?: string;
+  /** The preferred subdomain to provision for the team if not yet created */
+  subdomain: string;
+  /** The public url of an image representing the team */
+  avatarUrl?: string | null;
+  /** Details of the authentication provider being used */
+  authenticationProviders: {
+    /** The name of the authentication provider, eg "google" */
+    name: string;
+    /** External identifier of the authentication provider */
+    providerId: string;
+  }[];
+  /** The IP address of the incoming request */
+  ip: string;
+  /** Optional callback after team is created */
+  onSuccess?: (team: Team, transaction: Transaction) => Promise<void>;
+};
+
+async function teamCreator({
+  name,
+  domain,
+  subdomain,
+  avatarUrl,
+  authenticationProviders,
+  ip,
+  onSuccess,
+}: Props): Promise<Team> {
+  // If the service did not provide a logo/avatar then we attempt to generate
+  // one via ClearBit, or fallback to colored initials in worst case scenario
+  if (!avatarUrl) {
+    avatarUrl = await generateAvatarUrl({
+      name,
+      domain,
+      id: subdomain,
+    });
+  }
+
+  const team = await sequelize.transaction(async (transaction) => {
+    const team = await Team.create(
+      {
+        name,
+        avatarUrl,
+        authenticationProviders,
+      },
+      {
+        include: ["authenticationProviders"],
+        transaction,
+      }
+    );
+
+    await Event.create(
+      {
+        name: "teams.create",
+        teamId: team.id,
+        ip,
+      },
+      {
+        transaction,
+      }
+    );
+
+    if (onSuccess) {
+      await onSuccess(team, transaction);
+    }
```

Comment:
🥦 this allows follow-up actions to be wrapped into the same transaction. Kinda weird looking to me but I think it's sane

Reply:
Yea, it's a little weird – the way we typically do this now is just have the command accept require the transaction be passed in